[FSDP] Sharded Grad Scaler

sisilmehta2000 opened the pull request:
Summary: Adding in a shard aware grad scaler for FSDP+MixedPrecision support

Test Plan: Tests added

Differential Revision: D35988676

rohan-varma commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,90 @@
+
```
you can reference other files in test/distributed/fsdp to fix the lint here

rohan-varma commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,90 @@
+
+params = "cpu_offload,backward_prefetch,sharding_strategy,mixed_precision"
+cpu_offload_config = [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
+backward_prefetch_config = [BackwardPrefetch.BACKWARD_PRE, BackwardPrefetch.BACKWARD_POST, None]
```
In the interest of TTS, I'm pretty sure we can omit backward prefetch tests as I don't expect them to have implication for sharded grad scaling. cc @zhaojuanmao just in case I'm missing something

rohan-varma commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,90 @@
+                backward_prefetch=backward_prefetch,
+                sharding_strategy=sharding_strategy,
+                mixed_precision=mp,
+                enable_sharded_grad_scaler=True,
```
is there a reason we're testing sharded grad scaler when mixed precision is _not_ enabled?

sisilmehta2000 replied:
Yeah, I was using it as a baseline test to compare DDP vs FSDP (sharded grad scaler turned on and off). That way the test actually compares the params after the iterations. This was pretty useful the first few times I ran it to identify edge cases where there were bugs in the sharded grad scaler code.

rohan-varma replied:
Sounds good!

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+    return tensor.is_cuda or tensor.device.type in ("xla", "cpu")
+
+
+class _GeneralMultiDeviceReplicator(_MultiDeviceReplicator):
```
Add a docstring about what this class enables over `_MultiDeviceReplicator` to explain why we need it? Is it just to override the supported devices?

sisilmehta2000 replied:
Correct just to override the supported devices. Leaving comment here.

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+
+
+def _is_supported_device(tensor: torch.Tensor):
+    return tensor.is_cuda or tensor.device.type in ("xla", "cpu")
```
FSDP doesn't support XLA at the moment so we can possibly omit, though it might just be worth keeping for when we eventually do support XLA, I'll let you make the call

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+
+class ShardedGradScaler(GradScaler):
+    """
+    ShardedGradScaler helps performance gradient scaling in a shard aware manner. It extends
```
nit: helps perform

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+                        self._lazy_init_scale_growth_tracker(val.device)
+                    assert self._scale is not None
+                stash.append(_GeneralMultiDeviceReplicator(self._scale))
+                scaled_val = val * stash[0].get(val.device)
```
Trying to understand, for lists of tensors input, we can append to stash many times, but we always use stash[0] still?

sisilmehta2000 replied:
indent bug. fixed.

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+                try:
+                    assert tensor.device == expected_device, "grads must be on the same device"
+                except AssertionError:
+                    logging.error("tensor device is %s and expected device is %s" % (tensor.device, expected_device))
```
seems a bit roundabout to try, raise an assert, and then catch the assert to warn? Could we just warn directly?

But also, should we just throw here because grads on different device than params seems unexpected, it would have an issue in `opt.step()`?

sisilmehta2000 replied:
Yeah raising error here

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+                # TODO faster logic?
+                if torch.isinf(tensor).any().item() is True or torch.isnan(tensor).any().item() is True:  # type: ignore
+                    found_inf.data = torch.tensor([1.0])
+                    break
```
this means some grads might be scaled and some might not be? I guess that is fine, because when we find inf, opt.step() will be a noop, and eventually we just won't use these grads basically right?

sisilmehta2000 replied:
Yup correct. Once we find inf/nan, opt.step is skipped.

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+
+        # Make sure that the calls are done before moving out.
+        if last_handle:
+            last_handle.wait()
```
hmm, i guess this is sort of using an assumption of NCCL IIUC, i.e. if we submit comm ops on the same stream on the same nccl pg (which is what we do here) then nccl takes care to run those sequentially currently I believe, even if they are not data dependent ops. But I guess to be more backend agnostic or robust, it might be worth to collect all the futures (shouldn't be too many as they are per device) and wait them all with `torch.futures.wait_all`

but this can be done as a follow up, I think currently this should be fine

sisilmehta2000 replied:
done.

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+        # Google says mypy struggles with defaultdicts type annotations.
+        per_device_and_dtype_grads = defaultdict(lambda: defaultdict(list))  # type: ignore[var-annotated]
+        with torch.no_grad():
+            for group in optimizer.param_groups:
```
When FSDP supports optimizer param groups, we should add the relevant tests to sharded grad scaler as well.

sisilmehta2000 replied:
Leaving a TODO here for that.

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+            elif isinstance(val, abc.Iterable):
+                iterable = map(apply_scale, val)
+                if isinstance(val, (list, tuple)):
+                    return type(val)(iterable)
```
Can we have tests where a loss that is a list of tensors is scaled?

sisilmehta2000 replied:
added

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,329 @@
+        for v in optimizer_state["found_inf_per_device"].values():
+            if v.device.type == "cpu":
+                v_on_cuda = v.cuda()
+                last_handle = dist.all_reduce(v_on_cuda, async_op=True, group=self.process_group)
```
Not sure if I completely follow this or I'm missing something, but doesn't this _not_ update the current `v` in optimizer_state[found_inf_per_device]?

v_on_cuda should be an entirely different tensor than v, and we pass this into all_reduce. Then we do v_on_cuda.cpu(), but I'm not sure what that does, it creates a CPU tensor that we don't assign?

Trying to figure out if the intention here is to convert the current `v` inplace and make sure `v` updates it, but if that is the case based on my understanding this might not be happening.

sisilmehta2000 replied:
Good catch. Fixed.

rohan-varma reviewed (commented):
Looks great to me overall, thanks for working on this! Have a few comments / questions, and will also wait for the additional tests you mentioned as TODO. Thanks for putting this together so quickly!

sisilmehta2000 commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,153 @@
+)
+
+
+if not dist.is_available():
```
Need to move this check into the ditributed tests on line 115

rohan-varma commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,153 @@
+subtest_name = functools.partial(subtest_name, test_name_mapping)
+
+
+class TestShardGradScaler(TestCase):
```
These tests don't really seem to depend on shards / multiple ranks? Should we file an issue to test this against traditional grad scaler and verify we have parity?

sisilmehta2000 replied:
They dont directly depend on shard/multiple ranks. GradScaler.py has a check where if cuda is not available, it disables GradScaling. https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L116

rohan-varma commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -0,0 +1,153 @@
+        mp = MixedPrecision(
+            param_dtype=torch.float16,
+            reduce_dtype=torch.float16,
+            buffer_dtype=torch.float16,
```
As a follow up, can we also add bfloat16 tests. bfloat16 is hardware conditional, there are some logic in test_fsdp_mixed_precision to make sure we only run it where appropriate.

sisilmehta2000 replied:
bfloat16 is tested in test_mixed_precision.py above.

rohan-varma reviewed (approved):
LGTM, thanks a bunch for working on this and thank you for the excellent testing coverage!

Please fix typing and CI issues before landing. Also had some comments / suggestions around better testing, but feel like they can be addressed as follow up tasks (if you decide not to address them in this PR, can you file follow up issues?)

rohan-varma commented on torch/distributed/fsdp/sharded_grad_scaler.py:
```diff
@@ -0,0 +1,332 @@
+    """
+    def __init__(
+        self,
+        init_scale: float = 2.0 ** 16,
```
Do we feel like these defaults are well-tuned or expect the user have to tune them for their use case?

sisilmehta2000 replied:
I used the ones from the Fairscale implementation. I think will need a little experiment across different types of models to change the defaults.

Commit:
[FSDP] Sharded Grad Scaler (#76918)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76918

Adding in a shard aware grad scaler for FSDP+MixedPrecision support

Test Plan: Tests added

Reviewed By: rohan-varma

Differential Revision: D35988676

fbshipit-source-id: 67a59cafafa9d5366c9ef1df22fbca3706ba8d2d
